Api

Tavbulatov opened the pull request.

Tavbulatov committed:
add service octokit, controlerr gist, table gists space admin

Tavbulatov committed:
add model gist and locales syntax

VidgarVii commented on .gitignore:
```diff
@@ -38,3 +38,5 @@
 yarn-debug.log*
 .yarn-integrity
+
+.env
```
в конце файла надо отставлять 1 пустую строчку. Это указатель на конец файла. Сейчас система думает, что происходит обрыв файла. Даже гитхаб ругается. No new lina at end of file

VidgarVii commented on Gemfile:
```diff
@@ -5,15 +5,21 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+# Rubocop
 gem 'rubocop', require: false
 gem 'rubocop-performance', '~> 1.14.3', require: false
 gem 'rubocop-rails', '~> 2.15.2', require: false
```
данные гемы используют в разработке. Соответственно следует их положить в группу development

VidgarVii commented on Gemfile:
```diff
@@ -5,15 +5,21 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+gem 'octokit'
+
+# env - гем для чтения файлов env формата
+gem 'dotenv-rails', groups: %i[development test]
```
группы development test объявлены ниже. Пиши гем там. Не стоит устраивать бардак в этом файле

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -17,6 +17,20 @@ def update
     end
   end
 
+  def gist
```
Gist это ресурс. Значит согласно rails way и REST для создания ресурса  и записи в БД объекта Gist должен быть отдельный контроллер с экшеном create

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -17,6 +17,20 @@ def update
+
+    gist = Gist.new(user: current_user, question: @test_passage.current_question, url: result.url)
+
+    flash = if gist.save
```
данные для записи в базу гиста берутся не из формы юзера. Значит проверять следует только себя. Если Гист в базе сохранён, то это уже баг системы. И максимум, что юзер может увидеть, это что какой-то баг есть. А ошибки валидации видеть не к чему.
я бы сразу использовал create!
Gist.create!(...)
А вот что стоит проверять это внешний фактор. Т.е. в данном случаи это результат работы сервиса. Это создания гист в гитхаб. Если гист не будет создан в гитхаб, опять же можно показать флеш сообщение, о том, что что-то пошло не так. И по логам уже выяснять. И если гист будет создан в гитхаб, только тогда можно будет писать что-то в базу

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,25 @@
+    @client.create_gist(gist_params)
+  end
+
+  def gist_params
```
стоит разобраться в интерфейсе текущего класса. Я не уверен, что все методы являются публичные

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,25 @@
+
+  def initialize(question, client = nil)
+    @question = question
+    @client = client || Octokit::Client.new(access_token: ENV.fetch('GITHUB_TOKEN', nil))
```
установка "нил" в аргументах - это возможно не лучшее решение
я бы для такого случая порекомендовал вариант установки дефолта, к примеру, такой:
```ruby
def initialize(question, client = default_client)
  ...
  @client = client
end
```
далее можно определить в методе `default_client` "дефолтного клиента"

```ruby
private

def default_client
  Octokit::Client.new(access_token: GITHUB_GIST_TOKEN)
end
```

Tavbulatov committed:
A lot of work

add gist - controller, implementation of a separate action for creating a gist, moved gems to development group, fix service style and add method defaul_client, fix localisation , add routes gists

Tavbulatov committed:
fixbug controller gists

VidgarVii commented on app/services/gist_question_service.rb:
```diff
@@ -0,0 +1,29 @@
+
+  def call
+    @client.create_gist(gist_params)
+  end
```
Метод call должен вернуть такой объект, который содержит все нужные методы и данные. И хорошо если всё это будет прописано в явном виде. Сейчас же не понятно от куда берутся методы такие как html_url. 
Для создания такого объекта можно воспользоваться Struct 
Минус твоего подхода в том, что если придётся менять гем, то придётся исправлять код в разных местах проекта. А класс GistQuestionService служит своего рода адаптером над октокит.  Его задача не только создавать гист. Это он вообще делегирует гему октокит. А скрыть сложность создания гиста и предоставить общий интерфейс, который не зависит от хттп клиента. И если придётся менять этот хттп клиент, то рефакторигну подвергнется только сервис  GistQuestionService
Вот в контроллере не понятно от куда взялся метод url. Реализация клиента октокит не должен просачиваться в контроллер. Это лишние связи. 
Тут хорошо бы обернуть ответа клиента в какой-то объект и возвращать этот объект. Можно использовать Struct
например так
```ruby
Result = Struct.new(:html_url) do
  def success?
    html_url.present?
  end
end
```
И этот объект возвращать

VidgarVii commented on app/controllers/gists_controller.rb:
```diff
@@ -0,0 +1,17 @@
+            else
+              Gist.create!(user: current_user, question: test_passage.current_question, url: result.url)
+
+              { notice: t('.succes', url: result.url) }
```
ссылка должна ввести на сам гист. Ссылка должна быть кликабельна

Tavbulatov committed:
add clickable link in flash and fix services

Tavbulatov committed:
corrected name method and added html_safe for flash messages

VidgarVii commented on app/controllers/test_passages_controller.rb:
```diff
@@ -1,7 +1,7 @@
 
 class TestPassagesController < ApplicationController
-  before_action :set_test_passage, only: %i[result update show]
+  before_action :set_test_passage, only: %i[result update show gist]
```
тут остались лишние артефакты
